diff --git a/src/makernote_int.cpp b/src/makernote_int.cpp
index 20cd7629..6453586a 100644
--- a/src/makernote_int.cpp
+++ b/src/makernote_int.cpp
@@ -95,7 +95,7 @@ namespace Exiv2 {
         {
             std::string homedir;
             std::string inifile;
-#if defined(_MSC_VER) || defined(__MINGW__)
+#if (defined(_MSC_VER) || defined(__MINGW__)) && !(defined(WINAPI_FAMILY) && (WINAPI_FAMILY != WINAPI_FAMILY_DESKTOP_APP))
             char path[MAX_PATH];
             if (SUCCEEDED(SHGetFolderPathA(NULL, CSIDL_PROFILE, NULL, 0, path))) {
                 homedir = std::string(path);
